app/api.js

```diff
@@ -10,13 +12,69 @@ var statechartDefinitionSubscriptions = {};
 module.exports = function (simulation, db) {
   var api = {};
 
-  function createStatechartDefinition(req, res, scName) {
+  function createStatechartDefinitionWithTarball (req, res, scName) {
+    // What is happening here:
+    // 1 - request body has tar stream which goes into tar parser
+    // 2 - "on entry" section is storing each file on memory for inspection, validation etc.
+    // 3 - files goes into the simulation server
+
+    var files = {};
+
+    req.pipe(tar.Parse()).
+      on('entry', function (entry) {
+        var fileContents = '';
+
+        entry.on('data', function (c) {
+          fileContents += c.toString();
+        });
+
+        entry.on('end', function () {
+          files[entry.path] = { content: fileContents };
+        });
+      }).
+      on('end', function () {
+        console.log('all ended', Object.keys(files));
+
+        // TODO: Validate all .scxml files with async.eachSeries
+        // TODO: Abort stream parsing if there is validation error
+        // TODO: Pick index.js or first .scxml file as main
+        // TODO: Create each scxml file as a statechart so invoke can work
+        // TODO: Save statecharts to DB
+        // TODO: Broadcast each scxml change
+        var mainFile = files['index.scxml'];
```

Comment:
Remove files intermediate data structure. Only read out index.scxml for validation purposes.